### PR TITLE
Throw if listener tries to change headers after writing to the output stream

### PR DESCRIPTION
**Pull request by dlstucki**
## Description
Throw InvalidOperationException if listener attempts to change status code, description, or headers after writing to the output stream.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [X] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module). **It will throw instead of silently not doing what the user's code in the listener attempted to do.**
- [X] If applicable, the public code is properly documented.
- [X] Pull request includes test coverage for the included changes.
- [X] The code builds without any errors.

### CODE REVIEW

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs [hunk 1]
```diff
@@ -12,8 +12,9 @@ namespace Microsoft.Azure.Relay
     /// Represents a response to a request being handled by a <see cref="HybridConnectionListener"/> object.
     /// This is modeled after System.Net.HttpListenerResponse.
     /// </summary>
-    public sealed class RelayedHttpListenerResponse : IDisposable
+    public sealed class RelayedHttpListenerResponse : ITraceSource, IDisposable
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: >ITraceSource [](start = 54, length = 12)

This is new from the internal version of this same fix.

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs [hunk 2]
```diff
@@ -100,6 +103,8 @@ public string StatusDescription
             }
         }
 
+        TrackingContext ITraceSource.TrackingContext => this.Context.TrackingContext;
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: >TrackingContext ITraceSource.TrackingContext => this.Context.TrackingContext; [](start = 8, length = 77)

This is new from the internal version of this same fix.

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs [hunk 3]
```diff
@@ -149,11 +154,51 @@ void Dispose(bool disposing)
...
+            this.readOnly = true;
+        }
+
+        void CheckDisposedOrReadOnly()
         {
             if (this.disposed)
             {
-                throw RelayEventSource.Log.ThrowingException(new ObjectDisposedException(this.GetType().FullName), this);
+                string message = this.Context.TrackingContext.EnsureTrackableMessage(this.ToString());
+                throw RelayEventSource.Log.ThrowingException(new ObjectDisposedException(message), this);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: This is new from the internal version of this same fix.

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs [hunk 4]
```diff
@@ -149,11 +154,51 @@ void Dispose(bool disposing)
...
+            }
+
+            if (this.readOnly)
+            {
+                string message = this.Context.TrackingContext.EnsureTrackableMessage(SR.ResponseBodyStarted);
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: >string message = this.Context.TrackingContext.EnsureTrackableMessage(SR.ResponseBodyStarted); [](start = 16, length = 93)

This is new from the internal version of this same fix.

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs [hunk 5]
```diff
@@ -149,11 +154,51 @@ void Dispose(bool disposing)
...
+                throw RelayEventSource.Log.ThrowingException(new InvalidOperationException(message), this);
+            }
+        }
+
+        class ResponseWebHeaderCollection : WebHeaderCollection
```
Comment (bainian12345) [inline src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: >ResponseWebHeaderCollection [](start = 14, length = 27)

Looking at the API doc, I wonder if we need to also check for AddWithoutValidate() and Clear()

Comment (dlstucki) [reply src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: Clear can be added. AddWithoutValidate seems to be a protected non-virtual method.

---
In reply to: [276867326](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276867326) [](ancestors = 276867326)

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs [hunk 6]
```diff
@@ -149,11 +154,51 @@ void Dispose(bool disposing)
...
+        {
+            this.readOnly = true;
+        }
+
+        void CheckDisposedOrReadOnly()
         {
             if (this.disposed)
             {
-                throw RelayEventSource.Log.ThrowingException(new ObjectDisposedException(this.GetType().FullName), this);
+                string message = this.Context.TrackingContext.EnsureTrackableMessage(this.ToString());
```
Comment (bainian12345) [inline src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: >this.ToString() [](start = 85, length = 15)

Did you mean this.TrackingContext.ToString()?

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs [hunk 7]
```diff
@@ -149,11 +154,51 @@ void Dispose(bool disposing)
...
+                this.response = response;
+            }
+
+            public override void Add(string name, string value)
+            {
+                this.response.CheckDisposedOrReadOnly();
+                base.Add(name, value);
+            }
+
+            public override void Remove(string name)
```
Comment (bainian12345) [inline src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: >Remove(string name) [](start = 33, length = 19)

Just curious, with override like this what happens to the other overloads in the base class? Are those inaccessible now?

Comment (dlstucki) [reply src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: We can only intercept methods which are virtual. I've gone through the list and don't see any others which are obvious.

If the base class is well-designed then all other overloads should eventually result in one of these being invoked.

---
In reply to: [276870012](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276870012) [](ancestors = 276870012)

Comment (dlstucki) [reply src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: Another example:
The base class' `public string this[string name]`
is implemented by calling
`public virtual void Set(string name, string value)`

---
In reply to: [276871458](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276871458) [](ancestors = 276871458,276870012)

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs (hunk 5 above)
Comment (dlstucki) [reply src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: Clear added in iteration 2.

---
In reply to: [276867837](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276867837) [](ancestors = 276867837,276867326)

File: src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs (hunk 6 above)
Comment (dlstucki) [reply src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs]: Not exactly. It's customary to pass the type name in as the parameter to ObjectDisposedException (that's what the code had been doing previously). I'm keeping that plus tracking Id and address.

If I just used TrackingContext.ToString() then the ObjectDisposedException wouldn't contain the text `RelayedHttpListenerResponse` which I want to be in there.

---
In reply to: [276868207](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276868207) [](ancestors = 276868207)

**Pull request merged**